### ISSUE TITLE
Loading node/dashboard pages causes apache to consume large amounts of memory

### ISSUE THREAD

**Issue by exodusftw** [propyless/panopuppet#148]
in two different instances of Panopuppet - one with around 200 nodes, one with 30k+ - we see extreme wait times on the dashboard and nodes pages (upwards of 8-10 minutes) before the page is rendered.

Also - it appears during this time, apache consumes all available system memory (32GB)

**Comment by propyless** [comment]
Hi @exodusftw 

Been investigating a bit, and the reason panopuppet uses so much RAM is because I don't seem to utilize `yield` so much. (I should maybe do so...) I just use "return" which will load all the puppetdb querie results into ram ( the entire dataset) do lots of stuff to calculate x and y and then return that.
It might be saving serveral copies of the data itself in all the functions its calling and so it may not be as optimized as it could be.

I've done some benchmarking with a 9.5k node environment ( was difficult for me to reach 30k) don't have the hardware for it.. :(

but 

`
tot_resource: 0.099994
avg_resource: 0.172021
nodes: 0.242833
event_counts: 0.947424
all_nodes: 2.180118
reports: 29.675703
`
Here is the amount of seconds it took to get data from each result set.
Maybe I can do some work to go around the usage of reports since that seems to be the most expensive transaction.

It would be nice to know how long it takes for curl to get all the data from puppetdb's report endpoint.
Could you try that and get back to me with the results?

**Comment by propyless** [comment]
I've been doing some benchmarking and I've been able to get it down to these speeds.
The logic to get mismatching timestamps is the slowest part at the moment.

```
avg_resource: 0.109904
tot_resource: 0.11569
failed_nodes: 0.18939
nodes: 0.448737
noop_nodes: 0.620952
unreported_nodes: 3.076109
event_counts: 3.786209
changed_nodes: 6.049845
all_nodes: 6.233566
API Calls Total: 6.344626 
Data extraction and dictify total: 8.543716 
# Total nodes found
count - all_nodes: 20001
count - failed_nodes: 27
count - unreported_nodes: 8907
count - changed_nodes: 10832
count - noop_nodes: 23
count - mismatching_nodes: 1341
Total time for all ops: 14.888342
```

Removing mismatching timestamps yields these results
```
tot_resource: 0.093094
avg_resource: 0.114621
failed_nodes: 0.17559
nodes: 0.496458
noop_nodes: 0.618084
unreported_nodes: 2.945996
all_nodes: 4.389277
event_counts: 4.980502
changed_nodes: 5.892512
API Calls Total: 6.276579 
Data extraction and dictify total: 0.021835 
# Total nodes found
count - all_nodes: 20001
count - failed_nodes: 26
count - unreported_nodes: 8772
count - changed_nodes: 11177
count - noop_nodes: 26
Total time for all ops: 6.298414
```

There is an option to create a new pages for mismatching timestamps so thats its kept separate from the dashboard to yield faster load times. 

What do you think @exodusftw ?

**Comment by propyless** [comment]
Please checkout the branch `test_dash`
Its not entirely complete yet but it should you a rough idea of (if there are any improvements for you)

**Comment by exodusftw** [comment]
The odd thing i'm now seeing after pushing test_dash is that the metrics I see are fairly quick, but the rendering of the page that returns the result set is still taking upwards of 10 minutes.


avg_resource: 0.045493
tot_resource: 0.056408
failed_nodes: 1.265024
changed_nodes: 1.284045
unreported_nodes: 1.440665
event_counts: 2.182972
noop_nodes: 2.503633
all_nodes: 6.513316
API Calls Total: 6.560167
Data extraction and dictify total: 19.959999
# Total nodes found
count - all_nodes: 35369
count - changed_nodes: 420
Total time for all ops: 26.520166

**Comment by propyless** [comment]
Hey.

The majority of work he now been done in Dev_test_dash. I think it was
mentioned in a post further down. Do you have time to try that instead?

**Comment by propyless** [comment]
Also the statistics you posted show that it took around 40 seconds to
gather all info.

**Comment by exodusftw** [comment]
yes - the operations took around 40 seconds, but the page either takes about 10 minutes to return results, or simply times out and errors out that it can't connect to puppetdb.

I'll pull the dev_test_dash branch, deploy, and follow-up

**Comment by exodusftw** [comment]
with the dev_test_branch in on a 38k node env, I get the following

```shell
[Thu Jul 14 09:57:17.597765 2016] [:error] [pid 32067:tid 140468212344576] Request avg_resource: 0.040959
[Thu Jul 14 09:57:17.616859 2016] [:error] [pid 32067:tid 140468212344576] Request tot_resource: 0.018994
[Thu Jul 14 09:57:17.874761 2016] [:error] [pid 32067:tid 140467844937472] Request tot_nodes: 0.316579
[Thu Jul 14 09:57:18.147197 2016] [:error] [pid 32067:tid 140468195559168] Request failed_nodes: 0.589778
[Thu Jul 14 09:57:18.385783 2016] [:error] [pid 32067:tid 140468203951872] Request changed_nodes: 0.828766
[Thu Jul 14 09:57:18.395471 2016] [:error] [pid 32067:tid 140468220737280] Request unreported_nodes: 0.838898
[Thu Jul 14 09:57:20.801145 2016] [:error] [pid 32067:tid 140467836544768] Request recent_nodes: 3.146076
[Thu Jul 14 09:57:21.183164 2016] [:error] [pid 32067:tid 140467828152064] Request event_counts: 0.381306
[Thu Jul 14 09:57:21.693395 2016] [:error] [pid 32067:tid 140467819759360] Request recent_nodes_count: 0.891353
[Thu Jul 14 09:57:22.276846 2016] [:error] [pid 32067:tid 140468229129984] Request pending_nodes: 4.720535
[Thu Jul 14 15:09:39.134704 2016] [core:error] [pid 32070:tid 140468583134976] [client 1<some_ip>:63137] End of script output before headers: wsgi.py, referer: http://<some_node>/pano/nodes/
```

**Comment by propyless** [comment]
Hm interesting....

**Comment by exodusftw** [comment]
I'm still seeing issues with `can't connect` to PuppetDB on the Overview - page, but this PR (https://github.com/propyless/panopuppet/pull/155) resolved the issues with the Node page
